%\VignetteIndexEntry{Differential splicing and sQTL analysis in RNA-seq with 'DRIMSeq' package}
%\VignettePackage{DRIMSeq}
%\VignetteEngine{knitr::knitr}

\documentclass[11pt]{article}
\usepackage[utf8]{inputenc}


<<style-knitr, eval=TRUE, echo=FALSE, results="asis">>=
BiocStyle::latex()
@


\bioctitle[Differential splicing and sQTL analysis in RNA-seq with \Rpackage{DRIMSeq} package]{Dirichlet-multinomial framework for differential splicing and sQTL analysis in RNA-seq.}
%% also: \bioctitle{Title used for both header and title page}
%% or... \title{Title used for both header and title page}
\author{Malgorzata Nowicka\footnote{gosia.nowicka@uzh.ch}, Mark Robinson}


\begin{document}
\maketitle 
\noindent This vignette describes version \Sexpr{packageDescription("DRIMSeq")$Version} of the \Rpackage{DRIMSeq} package.
\tableofcontents


<<setup_knitr, include=FALSE, cache=FALSE>>=
library(knitr)
opts_chunk$set(cache = TRUE, tidy = TRUE, tidy.opts = list(blank = FALSE, width.cutoff = 60), out.width = "8cm", out.height = "8cm", fig.align = "center")
@



%------------------------------------------------------------------------------
%	Introduction
%------------------------------------------------------------------------------

\section{Overview of Dirichlet-multinomial model}





%------------------------------------------------------------------------------
%	Differential splicing analysis work-flow
%------------------------------------------------------------------------------

\section{Differential splicing analysis work-flow}

\subsection{Example data}

FASTQ files can be downloaded from ... (See Charlotte's paper)
Reference genome files from ...

\subsection{RNA-seq data alignment and counting}

Step depend on which counts you want to use:

Exonic bin counts from HTSeq/DEXSeq, featureCounts 

Transcript counts from kallisto, RSEM, BitSeq

(Reference to "Count-based differential expression...", pasilla package)


\subsection{Differential splicing analysis with \Rpackage{DM} package}

Assuming that you have a table with feature counts...

Load \Rpackage{DRIMSeq} package.

<<library>>=
library(DRIMSeq)
@


Create a \Rcode{dmDSdata} object that contains counts.

<<DSdmDSdata>>=

library(pasilla)

data_dir  <- system.file("extdata", package="pasilla")
count_files <- list.files(data_dir, pattern="fb.txt$", full.names=TRUE)
count_files

# Create a data frame with htseq counts
htseq_list <- lapply(1:length(count_files), function(i){
  # i = 1
  htseq <- read.table(count_files[i], header = FALSE, as.is = TRUE)
  colnames(htseq) <- c("group_id", gsub("fb.txt", "", strsplit(count_files[i], "extdata/")[[1]][2]))
  return(htseq)
})

htseq_counts <- Reduce(function(...) merge(..., by = "group_id", all=TRUE, sort = FALSE), htseq_list)
tail(htseq_counts)
htseq_counts <- htseq_counts[!grepl(pattern = "_", htseq_counts$group_id), ]

group_split <- limma::strsplit2(htseq_counts[, 1], ":")

d <- dmDSdata(counts = htseq_counts[, -1], gene_id = group_split[, 1], feature_id = group_split[, 2], sample_id = colnames(htseq_counts)[-1], group = gsub("[1-4]", "", colnames(htseq_counts)[-1]))

plotData(d)

# Use a subset of genes, which is defined in the following file
genes_subset = readLines(file.path(data_dir, "geneIDsinsubset.txt"))
d <- d[names(d) %in% genes_subset, ]

plotData(d)

@


Filter genes and transcripts with low expression.

<<DSdmFilter>>=
d <- dmFilter(d)
plotData(d)
@


Estimate dispersion.

<<DSdmDispersion>>=
d <- dmDispersion(d, BPPARAM = BiocParallel::MulticoreParam(workers = 2))
plotDispersion(d)
@


Estimate proportions.



<<DSdmFit, out.width = "16cm", fig.width = 16>>=
d <- dmFit(d)

gene_id <- names(d)[1]
plotFit(d, gene_id = gene_id, plot_type = "barplot")

@


Use likelihood ratio to test for differential splicing.


<<DSdmTest>>=
d <- dmTest(d)

plotTest(d)

res <- results(d)

@


<<DSdmLRT_plotFit, out.width = "16cm", fig.width = 16>>=
gene_id <- res$gene_id[1]
plotFit(d, gene_id = gene_id)
@




%------------------------------------------------------------------------------
%	sQTL analysis work-flow
%------------------------------------------------------------------------------


\section{sQTL analysis work-flow}

\subsection{Example data}

Data downloaded from

\subsection{Extracting the bi-allelic SNPs from CSV files}

Preprocessing steps on CSV files with genotypes


\subsection{sQTL analysis with \Rpackage{DM} package}

Assuming you have counts and bi-allelic genotypes...






\end{document}